### PR TITLE
HIVE-26285: Overwrite database metadata on original source in optimised failover.

### PR DESCRIPTION
**Pull request by hmangla98**

### CODE REVIEW

**Comment by achennagiri** [comment]
Hi @hmangla98,
Can you add some description to the ticket? I couldn't find any details in the JIRA as well.
Thanks.

**Comment by hmangla98** [comment]
> Hi @hmangla98, Can you add some description to the ticket? I couldn't find any details in the JIRA as well. Thanks.

Hey @achennagiri, In case of failover, there may be a situation that new alter database operations on original source were not replicated to original target. So, we need to overwrite those alter database events while replicating in reverse direction (i.e target --> source) just to keep db properties in sync.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 1]
```diff
@@ -703,6 +705,7 @@ private void createBuilderTask(List<Task<?>> rootTasks) {
   private int executeIncrementalLoad(long loadStartTime) throws Exception {
     // If replication policy is changed between previous and current repl load, then drop the tables
     // that are excluded in the new replication policy.
+    this.childTasks = (this.childTasks == null) ? new ArrayList<>() : this.childTasks;
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:708]: How much of this refactor needed for the specific change?
We might want to do it in a separate PR so we can backport the fix without the conflicts generated by the refactor when needed

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 2]
```diff
@@ -825,4 +836,15 @@ private int executeIncrementalLoad(long loadStartTime) throws Exception {
     LOG.info("REPL_INCREMENTAL_LOAD stage duration : {} ms", currentTimestamp - loadStartTime);
     return 0;
   }
+
+  private Database getSourceDbMetadata() throws IOException, SemanticException {
+    Path dbMetadata = new Path(work.dumpDirectory, EximUtil.METADATA_PATH_NAME);
+    BootstrapEventsIterator itr = new BootstrapEventsIterator(dbMetadata.toString(), work.dbNameToLoadIn,
+            true, conf, work.getMetricCollector());
+    assert itr.hasNext();
```
[the remark this answers is not in the document]
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:844]: Added semantic exception to fail replication in that case.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 3]
```diff
@@ -739,6 +741,23 @@ private int executeIncrementalLoad(long loadStartTime) throws Exception {
         LOG.info("Dropping table {} for optimised bootstarap", work.dbNameToLoadIn + "." + table);
         db.dropTable(work.dbNameToLoadIn + "." + table, true);
       }
+      Database sourceDb = getSourceDbMetadata();  //This sourceDb was the actual target prior to failover.
+      Map<String, String> sourceDbProps = sourceDb.getParameters();
+      List<String> replicationProps = MetaStoreUtils.getReplicationDbProps(); //Replication Props will be handled separately as part of preAckTask.
+      for (Map.Entry<String, String> currProp : targetDb.getParameters().entrySet()) {
+        String actualVal = sourceDbProps.get(currProp.getKey());
+        if (!replicationProps.contains(currProp.getKey()) && !currProp.getValue().equals(actualVal)) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:749]: can't we filter the map (targetDb.getParameters()) with replicationProps first to avoid the lookups?
````
    replicationProps.stream()
              .map(targetDb.getParameters()::get)
              .filter(Objects::isNull)
               ....
````

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 4]
```diff
@@ -825,4 +844,16 @@ private int executeIncrementalLoad(long loadStartTime) throws Exception {
     LOG.info("REPL_INCREMENTAL_LOAD stage duration : {} ms", currentTimestamp - loadStartTime);
     return 0;
   }
+
+  private Database getSourceDbMetadata() throws IOException, SemanticException {
+    Path dbMetadata = new Path(work.dumpDirectory, EximUtil.METADATA_PATH_NAME);
+    BootstrapEventsIterator itr = new BootstrapEventsIterator(dbMetadata.toString(), work.dbNameToLoadIn,
+            true, conf, work.getMetricCollector());
+    BootstrapEvent next = itr.next();
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:852]: should we check for NULL?

**Review by deniskuzZ — APPROVED** [review]
